### ISSUE TITLE
Plugins initialise twice and break in react wrapper

### ISSUE THREAD

**Issue by jonom** [katspaugh/wavesurfer.js#3731]
## Bug description

TLDR; I can't get the spectrogram plugin to work on the react wrapper. I haven't tested other types of plugins.

**Note:** I think this is an issue with https://github.com/katspaugh/wavesurfer-react rather than this repo, but there is no issues tab on the react repo so I'm posting here. Also, it might be able to be fixed in this repo by making plugins reusable - i.e., able to be re-initialised.

The issue seems to be that the plugin instance is mutated during initialisation, and if you attempt to initialise the plugin a second time it will cause an error. The problem seems to be in the [useEffect](https://github.com/katspaugh/wavesurfer-react/blob/main/src/index.tsx#L56) hook which destroys and recreates the wavesurfer instance. In subsequent runs of the hook a new ws instance would get created but it would receive already-initialised plugin instances.

I found that if I memoised all of the props being passed to WavesurferPlayer I could get the plugin to work when building for production at least, but when running locally it still breaks since React throws in that 'helpful' extra render.

## Environment
 - Browser: Chrome

## Minimal code snippet

```jsx
import { useCallback, useRef, useState } from "react";
import WavesurferPlayer from "@wavesurfer/react";
import Spectrogram from "wavesurfer.js/dist/plugins/spectrogram.esm.js";
import createColormap from "colormap";

function AudioPlayer({ audioUrl }) {
  const [wavesurfer, setWavesurfer] = useState(null);
  const [isPlaying, setIsPlaying] = useState(false);
  const plugins = useRef(null);
  if (plugins.current === null) {
    plugins.current = [
      Spectrogram.create({
        labels: true,
        height: 200,
        splitChannels: false,
        colorMap: createColormap({
          colormap: "jet",
          nshades: 256,
          format: "float",
        }),
      }),
    ];
  }

  const onReady = useCallback((ws) => {
    setWavesurfer(ws);
    setIsPlaying(false);
  }, []);

  const onPlayPause = useCallback(() => {
    wavesurfer && wavesurfer.playPause();
  }, [wavesurfer]);

  const onPlay = useCallback(() => {
    setIsPlaying(true);
  }, []);

  const onPause = useCallback(() => {
    setIsPlaying(false);
  }, []);

  return plugins.current ? (
    <>
      <WavesurferPlayer
        height={100}
        waveColor="violet"
        url={audioUrl}
        onReady={onReady}
        onPlay={onPlay}
        onPause={onPause}
        plugins={plugins.current}
      />

      <button onClick={onPlayPause}>{isPlaying ? "Pause" : "Play"}</button>
    </>
  ) : null;
}

export default AudioPlayer;

```

## Expected result

A player with a spectrogram.

## Obtained result

TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.

**Comment by jonom** [comment]
FWIW I also tried adding the plugin in `onInit` and `onReady` but it didn't work - I got a regular player without the plugin rendered.

```jsx
  const onReady = useCallback((ws) => {
    setWavesurfer(ws);
    setIsPlaying(false);
    ws.registerPlugin(
      Spectrogram.create({
        labels: true,
        height: 200,
        splitChannels: false,
        colorMap: createColormap({
          colormap: "jet",
          nshades: 256,
          format: "float",
        }),
      })
    );
  }, []);
```

**Comment by brianlaw033** [comment]
I came across this issue too, and looks like its react's strict mode that leads to useEffect being ran twice
The solution for me is to run `destroy()` on unmount

**Comment by katspaugh** [comment]
Good catch! Looks like this ref usage leads to different lifecycles between a wavesurfer instance and a plugin instance.

As a workaround, try using the `useWavesurfer` hook instead of the component. See an example here: https://wavesurfer.xyz/examples/?react.js

**Comment by jonom** [comment]
Thanks for your replies! Back on this project today and I tried using the `useWavesurfer` hook but it has the same result because the component uses those hooks under the hood. Note that I tried with the `Timeline` plugin instead of `Spectrogram` and that worked fine, so I guess this issue is plugin-specific.

My solution for now is to modify `useWavesurferInstance()` to accept a getPlugins function:

```jsx
    const ws = WaveSurfer.create({
      ...options,
      container: containerRef.current,
      plugins: options.getPlugins ? options.getPlugins() : options.plugins,
    });
```

```jsx
  const wavesurferOptions = useMemo(
    () => ({
      container: containerRef,
      url: audioUrl,
      getPlugins: () => [
        Spectrogram.create(),
      ],
    }),
    [audioUrl]
  );

  const { wavesurfer, isPlaying, currentTime } =
    useWavesurfer(wavesurferOptions);
```

This way a fresh plugin instance gets created each time the player initialises.

**Comment by robotastic** [comment]
@jonom - thanks for that snippet, it got things working for me.

**Comment by katspaugh** [comment]
> Thanks for your replies! Back on this project today and I tried using the useWavesurfer hook but it has the same result

I've just re-read your initial issue description and realized you already narrowed it down to the extra render React does in dev mode, so both the hook and the component have the same problem.

I guess one workaround would be to init/destroy plugins in `onInit` and `onDestroy` but maybe there's some clever way to handle this in the lib itself.

**Comment by vlad-solomon** [comment]
@brianlaw033 Yup, I've come to the same conclusion as you and @jonom . It's React's Strict Mode that causes this. What I'm wondering is what do you mean by this? 

> The solution for me is to run `destroy()` on unmount

This solved your problem? Because I'm trying to do the same thing and it doesn't work. I don't want to disable strict mode because that catches plenty of bugs. Is there something I'm missing? Here's my code:

```jsx
export default function Track({ index, audio }) {
    const video = useVideoStore((state) => state.video);
    const [wavesurfer, setWavesurfer] = useState(null);

    const onReady = (ws) => {
        setWavesurfer(ws);
    };

    useEffect(() => {
        return () => {
            if (wavesurfer) {
                wavesurfer.stop();
                wavesurfer.destroy();
            }
        };
    }, [video, wavesurfer]);

    return (
        <div className="h-10 border-b border-coal-800 last:border-0">
            <WavesurferPlayer
                interact={false}
                height={40}
                waveColor={waveColor}
                progressColor={progressColor}
                url={audio}
                autoplay={true}
                onReady={onReady}
            />
        </div>
    );
}
```

**Comment by sebbean** [comment]
> Thanks for your replies! Back on this project today and I tried using the `useWavesurfer` hook but it has the same result because the component uses those hooks under the hood. Note that I tried with the `Timeline` plugin instead of `Spectrogram` and that worked fine, so I guess this issue is plugin-specific.
> 
> My solution for now is to modify `useWavesurferInstance()` to accept a getPlugins function:
> 
>     const ws = WaveSurfer.create({
>       ...options,
>       container: containerRef.current,
>       plugins: options.getPlugins ? options.getPlugins() : options.plugins,
>     });
>   const wavesurferOptions = useMemo(
>     () => ({
>       container: containerRef,
>       url: audioUrl,
>       getPlugins: () => [
>         Spectrogram.create(),
>       ],
>     }),
>     [audioUrl]
>   );
> 
>   const { wavesurfer, isPlaying, currentTime } =
>     useWavesurfer(wavesurferOptions);
> This way a fresh plugin instance gets created each time the player initialises.

@jonom any chance you have this on a fork / branch i could reference?

**Comment by jonom** [comment]
@sebbean the code I ended up coming up with [is here](https://github.com/orcasound/seastats-dashboard/blob/main/src/components/AudioPlayer.jsx), but the important part is a hook that creates the WS instance, and destroys it and creates a new one if any dependencies change. That way you can avoid the double plugin initialization.

Here's a simplified example:

```jsx
import { useEffect, useRef, useState } from "react";
import Spectrogram from "wavesurfer.js/dist/plugins/spectrogram.esm.js";
import WaveSurfer from "wavesurfer.js";

function AudioPlayer({
  audioUrl,
  fftSamples = 4096,
  frequencyMin = 0,
  frequencyMax = 8000,
}) {
  const containerRef = useRef(null);
  const [wavesurfer, setWavesurfer] = useState(null);
  const [playing, setPlaying] = useState(false);

  useEffect(() => {
    if (!containerRef.current) return;

    const ws = WaveSurfer.create({
      container: containerRef.current,
      height: 0,
      url: audioUrl,
      sampleRate: 88200,
      plugins: [
        Spectrogram.create({
          labels: true,
          height: 200,
          splitChannels: false,
          fftSamples,
          frequencyMin,
          frequencyMax,
        })],
    });

    setWavesurfer(ws);

    const unsubscribe = [
      ws.on("play", () => {
        setPlaying(true);
      }),
      ws.on("pause", () => {
        setPlaying(false);
      }),
    ];

    return () => {
      unsubscribe.forEach((fn) => fn());
      ws.destroy();
    };
  }, [
    audioUrl,
    fftSamples,
    frequencyMax,
    frequencyMin,
  ]);

  return (
    <div>
      <div className="player" ref={containerRef} />
      <div className="controls">
        <button onClick={() => wavesurfer.playPause()}>
          {playing ? "Pause" : "Play"}
        </button>
      </div>
    </div>
  );
}

export default AudioPlayer;
```

**Comment by sebbean** [comment]
ty @jonom !
